### PR TITLE
RavenDB-21484: Current behavior matches the Lucene behavior even though it is suboptimal.

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-21484

### Additional description

It is a technicality but at the select `select id() as Id, Title, Url, Category` there is no `Content` field. Corax was not considering the highlight field and therefore didn't get it from the storage.

### Type of change
- Bug fix

### How risky is the change?
- Low 

### Backward compatibility
- Non breaking change

### CODE REVIEW

File: src/Raven.Server/Documents/Queries/Highlightings/HighlightingField.cs [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 namespace Raven.Server.Documents.Queries.Highlightings
 {
-    public sealed class HighlightingField
```
Comment (maciejaszyk) [inline src/Raven.Server/Documents/Queries/Highlightings/HighlightingField.cs]: Can we project a field Content (without highlighting) and highlight Content at the same time?

File: src/Raven.Server/Documents/Queries/FieldsToFetch.cs [hunk 2]
```diff
@@ -258,6 +258,22 @@ private static void ThrowInvalidFetchAllStoredDocuments()
                 result[key] = val;
             }
 
+            if (metadata.HasHighlightings)
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/FieldsToFetch.cs]: Can we support highlights without projecting the field?

File: src/Raven.Server/Documents/Queries/Highlightings/HighlightingField.cs (hunk 1 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/Queries/Highlightings/HighlightingField.cs]: Yes. Because we essentially project it only if it wasnt projected before. Having said that, I really dislike the solution but that is how Lucene works.

File: src/Raven.Server/Documents/Queries/FieldsToFetch.cs (hunk 2 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/Queries/FieldsToFetch.cs]: If the field is a projection, nope. The big question is if we can avoid send it to the client which I added an issue to handle. https://issues.hibernatingrhinos.com/issue/RavenDB-21496/Corax-shouldnt-return-the-highlighted-field-if-it-is-not-selected-in-the-projection.

Comment (arekpalinski) [reply src/Raven.Server/Documents/Queries/FieldsToFetch.cs]: This adds a highlighted field to be also returned in projection (even if not requested). Effectively this will also change the behavior of Lucene queries (previously we didn't return the field in projection, now it will).

Can we retrieve that field during query execution (since we need the value to handle highlighting) but skip returning it on the projection?

Comment (redknightlois) [reply src/Raven.Server/Documents/Queries/FieldsToFetch.cs]: This is when explicitly required:
image
This is when not:
image

How did you trigger the difference? Because `QueryMetadata.HasHighlightings` should be `false` if we are not asking for including the highlightings, it may be a query parsing issue.

Comment (arekpalinski) [reply src/Raven.Server/Documents/Queries/FieldsToFetch.cs]: The issue is that even if we _don't_ request `Content` field in projection we will still return it. The following query:

```
from index 'MultiSearch' select id() as Id, Title, Url, Category include highlight(Content,128,1,$p0)
{"p0":{"PreTags":["<span class='search-result-highlight'>"],"PostTags":["</span>"]}}
```

didn't return `Content` field previously:

image

but with this change it will return it (on Lucene and Corax):

image


So my question is: can we retrieve that field internally for Corax in order to do highlighting but do not return in in projection results since it was not requested?

File: src/Raven.Server/Documents/Queries/FieldsToFetch.cs [hunk 3]
```diff
@@ -36,13 +36,13 @@ public sealed class FieldsToFetch
 
         public readonly ProjectionOptions Projection;
 
-        public FieldsToFetch(IndexQueryServerSide query, IndexDefinitionBaseServerSide indexDefinition, IndexType indexType, SearchEngineType searchEngineType = SearchEngineType.None)
+        public FieldsToFetch(IndexQueryServerSide query, IndexDefinitionBaseServerSide indexDefinition, IndexType indexType, SearchEngineType searchEngineType)
         {
             Projection = new ProjectionOptions(query);
-            Fields = GetFieldsToFetch(query.Metadata, query.ProjectionBehavior, indexDefinition, searchEngineType, indexType, out AnyExtractableFromIndex, out bool extractAllStoredFields, out SingleBodyOrMethodWithNoAlias, out AnyTimeSeries);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Queries/FieldsToFetch.cs:42]: So now we don't need `searchEngineType` at all? 

I was thinking about handling it this way (if possible):

> can we retrieve that field internally for Corax in order to do highlighting but do not return in in projection results since it was not requested?

So for that we'd need to know the engine type. 

https://github.com/ravendb/ravendb/pull/17435#discussion_r1343587184

File: src/Raven.Server/Documents/Queries/FieldsToFetch.cs [hunk 4]
```diff
@@ -217,11 +215,43 @@ private static void ThrowInvalidFetchAllStoredDocuments()
             throw new InvalidOperationException("Cannot fetch all stored path from a nested method");
         }
 
+        public static Dictionary<string, FieldToFetch> GetFieldsToHighlight(
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Queries/FieldsToFetch.cs]: I don't see any usage of this method

File: src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs [hunk 5]
```diff
@@ -136,12 +136,12 @@ protected void FinishDocumentSetup(Document doc, ref RetrieverInput retrieverInp
 
         protected abstract DynamicJsonValue GetCounterRaw(string docId, string name);
 
-        protected (Document Document, List<Document> List) GetProjection(ref RetrieverInput retrieverInput, string lowerId, CancellationToken token)
+        protected (Document Document, List<Document> List) GetProjection(ref RetrieverInput retrieverInput, FieldsToFetch fieldsToFetch, string lowerId, CancellationToken token)
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:139]: Why `fieldsToFetch` is passed as parameter now? I see that in all usages we pass `FieldsToFetch` field.

File: test/SlowTests/Issues/RavenDB_21525.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+public class RavenDB_21525 : RavenTestBase
+{
+    public RavenDB_21525(ITestOutputHelper output) : base(output)
+    {
+    }
+
+
+    [RavenTheory(RavenTestCategory.Corax)]
+    [RavenData(SearchEngineMode = RavenSearchEngineMode.All)]
+    public async Task OrderByWithPaginationVsUnordered(Options options, int size = 10000)
```
Comment (arekpalinski) [inline test/SlowTests/Issues/RavenDB_21525.cs]: Why do we have a test from different PR https://github.com/ravendb/ravendb/pull/17486 ?

Comment (arekpalinski) [reply test/SlowTests/Issues/RavenDB_21525.cs]: Where is the test for the issue solved here - RavenDB-21484?

**Review by arekpalinski — COMMENTED** [review]
Do we need to have changes from https://github.com/ravendb/ravendb/pull/17486 here as well?

**Comment by arekpalinski** [comment]
https://github.com/ravendb/ravendb/pull/17486 got merged. Please rebase

**Comment by redknightlois** [comment]
This is still work in progress.

File: src/Raven.Server/Documents/Queries/FieldsToFetch.cs (hunk 3 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/Queries/FieldsToFetch.cs:42]: No, that would cause the returning of the projection result, which we don't want. It is proving much more difficult than expected to obtain the same behavior.

File: src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs (hunk 5 above)
Comment (redknightlois) [reply src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:139]: To unify and extract those methods to be able to be used for highlighting as a special codepath.

File: test/SlowTests/Issues/RavenDB_21525.cs (hunk 6 above)
Comment (redknightlois) [reply test/SlowTests/Issues/RavenDB_21525.cs]: That tests is able to showcase both this and RavenDB-21525

Comment (arekpalinski) [reply test/SlowTests/Issues/RavenDB_21525.cs]: How it could showcase this case while there is no highlighting involved here?

**Comment by arekpalinski** [comment]
> This is still work in progress.

Oh, marking it as draft PR then. Please let me know when it's ready for review

**Comment by arekpalinski** [comment]
Merged in https://github.com/ravendb/ravendb/pull/17571
